File: miniapp/src/browser/keys.ts
// What the operator's keyboard and mouse become on the page while they drive.
//
// Printable text does not go as key presses: it goes as `text` (the daemon's `Input.insertText`), from
// the viewer's hidden field, which is the one place a browser hands over composed text — an IME's
// finished word, a phone keyboard's autocorrection, dictation, a paste. What is sent as a key is what
// has no text: Enter, Tab, Backspace, the arrows, Escape, and anything held with Ctrl, Alt or Meta, so
// Ctrl+A selects on the page and Ctrl+Enter submits a form.
//
// A few chords are the viewer's rather than the page's, the ones a person uses on a browser without
// thinking: the address bar, reload, back and forward. They would otherwise go to the app's own tab.

import type { InputMessage, Mods } from "./protocol";

/** CDP's modifier bits, which the daemon passes on as they are. */
export const MOD = { alt: 1, ctrl: 2, meta: 4, shift: 8 } as const;

type KeyLike = Pick<KeyboardEvent, "key" | "code" | "keyCode" | "altKey" | "ctrlKey" | "metaKey" | "shiftKey">;
type MouseLike = Pick<MouseEvent, "altKey" | "ctrlKey" | "metaKey" | "shiftKey">;

export function modsOf(e: MouseLike): Mods {
  return (e.altKey ? MOD.alt : 0) | (e.ctrlKey ? MOD.ctrl : 0) | (e.metaKey ? MOD.meta : 0) | (e.shiftKey ? MOD.shift : 0);
}

/** Keys the page gets as presses. A letter alone is text; the same letter with Ctrl is a key. */
const NAMED = new Set([
  "Enter", "Tab", "Backspace", "Delete", "Escape", "ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight",
  "Home", "End", "PageUp", "PageDown", "Insert",
  "F1", "F2", "F3", "F4", "F5", "F6", "F7", "F8", "F9", "F10", "F11", "F12",
]);

/** Windows virtual key codes for the named keys, which CDP wants beside the key's name. */
const KEY_CODES: Record<string, number> = {
  Backspace: 8, Tab: 9, Enter: 13, Escape: 27, " ": 32, PageUp: 33, PageDown: 34, End: 35, Home: 36,
  ArrowLeft: 37, ArrowUp: 38, ArrowRight: 39, ArrowDown: 40, Insert: 45, Delete: 46,
};

/** Whether a key press goes to the page as a key (true) or is left to arrive as text (false). */
export function isKeyPress(e: KeyLike): boolean {
  if (e.key === "Dead" || e.key === "Process" || e.key === "Unidentified") return false;
  if (["Shift", "Control", "Alt", "Meta", "CapsLock", "AltGraph"].includes(e.key)) return false;
  if (NAMED.has(e.key)) return true;
  // A character with Ctrl or Meta is a shortcut on the page (Ctrl+A); with Alt alone it is often a
  // character on its own (Option+e on a Mac, AltGr on a European layout), so it stays text.
  return (e.ctrlKey || e.metaKey) && e.key.length === 1;
}

function keyCode(e: KeyLike): number {
  if (KEY_CODES[e.key] !== undefined) return KEY_CODES[e.key];
  if (/^F([1-9]|1[0-2])$/.test(e.key)) return 111 + Number(e.key.slice(1));
  if (e.key.length === 1) {
    const upper = e.key.toUpperCase();
    if (/[A-Z0-9]/.test(upper)) return upper.charCodeAt(0);
  }
  return e.keyCode || 0;
}

/** The INPUT for one key going down or up. `text` rides with Enter, which is how a page sees a newline typed. */
export function keyInput(e: KeyLike, type: "down" | "up"): Extract<InputMessage, { t: "key" }> {
  const text = type === "down" && e.key === "Enter" && !e.ctrlKey && !e.metaKey && !e.altKey ? "\r" : undefined;
  return { t: "key", type, key: e.key, code: e.code, key_code: keyCode(e), ...(text ? { text } : {}), mods: modsOf(e) };
}

/** A named key pressed and released, from the phone's row of keys. */
export function tapKey(key: string): Extract<InputMessage, { t: "key" }>[] {
  const code = key.startsWith("Arrow") || NAMED.has(key) ? key : "";
  const e = { key, code, keyCode: 0, altKey: false, ctrlKey: false, metaKey: false, shiftKey: false };
  return [keyInput(e, "down"), keyInput(e, "up")];
}

export function mouseButton(button: number): "left" | "middle" | "right" {
  return button === 1 ? "middle" : button === 2 ? "right" : "left";
}

/** The viewer's own chords, caught before the page or the app's tab gets them. */
export type ViewerChord = "address" | "reload" | "back" | "forward" | "swallow";

export function viewerChord(e: KeyLike, mac: boolean): ViewerChord | null {
  const primary = mac ? e.metaKey && !e.ctrlKey : e.ctrlKey && !e.metaKey;
  if (primary && !e.altKey) {
    const k = e.key.toLowerCase();
    if (k === "l") return "address";
    if (k === "r") return "reload";
    if (e.key === "[") return "back";
    if (e.key === "]") return "forward";
    // New tab and close tab would act on the app's own window; the viewer has no tabs of its own to
    // open or close for a person yet, so the chord is simply kept from leaving.
    if (k === "t" || k === "w") return "swallow";
  }
  if (!mac && e.altKey && !e.ctrlKey && !e.metaKey) {
    if (e.key === "ArrowLeft") return "back";
    if (e.key === "ArrowRight") return "forward";
  }
  if (e.key === "F5" && !e.altKey) return "reload";
  return null;
}

/**
 * Escape belongs to the page first — a menu, a modal, a search field all close with it — so the viewer
 * lets go of the keyboard only on a second press within this long.
 */
export const ESCAPE_TWICE_MS = 600;

File: browserd/internal/page/keys.go
package page

import (
	"fmt"
	"strings"
	"unicode"
	"unicode/utf8"
)

// key is what CDP needs to press one key.
type key struct {
	Key     string
	Code    string
	KeyCode int
	Text    string
}

// named are the keys the contract names.
var named = map[string]key{
	"Enter":      {"Enter", "Enter", 13, "\r"},
	"Tab":        {"Tab", "Tab", 9, ""},
	"Escape":     {"Escape", "Escape", 27, ""},
	"Backspace":  {"Backspace", "Backspace", 8, ""},
	"Delete":     {"Delete", "Delete", 46, ""},
	"Space":      {" ", "Space", 32, " "},
	"ArrowUp":    {"ArrowUp", "ArrowUp", 38, ""},
	"ArrowDown":  {"ArrowDown", "ArrowDown", 40, ""},
	"ArrowLeft":  {"ArrowLeft", "ArrowLeft", 37, ""},
	"ArrowRight": {"ArrowRight", "ArrowRight", 39, ""},
	"Home":       {"Home", "Home", 36, ""},
	"End":        {"End", "End", 35, ""},
	"PageUp":     {"PageUp", "PageUp", 33, ""},
	"PageDown":   {"PageDown", "PageDown", 34, ""},
}

// Modifier bits, as CDP counts them, and the keys that hold them.
var modifiers = map[string]struct {
	bit int
	key key
}{
	"Alt":   {1, key{"Alt", "AltLeft", 18, ""}},
	"Ctrl":  {2, key{"Control", "ControlLeft", 17, ""}},
	"Meta":  {4, key{"Meta", "MetaLeft", 91, ""}},
	"Shift": {8, key{"Shift", "ShiftLeft", 16, ""}},
}

// chord is one press: the modifiers held and the key.
type chord struct {
	mods    []string
	modBits int
	key     key
}

// parseKeys reads "Enter", "Ctrl+A", "Shift+Tab", "a", "F5".
func parseKeys(s string) (chord, error) {
	if s == "" || len(s) > 64 {
		return chord{}, fmt.Errorf("keys must name a key, such as Enter, Tab, Ctrl+A")
	}
	parts := strings.Split(s, "+")
	// "Ctrl++" is Ctrl and the plus key.
	if strings.HasSuffix(s, "++") {
		parts = append(strings.Split(strings.TrimSuffix(s, "++"), "+"), "+")
	}
	var c chord
	for _, m := range parts[:len(parts)-1] {
		mod, ok := modifiers[m]
		if !ok {
			return chord{}, fmt.Errorf("unknown modifier %q; use Ctrl, Shift, Alt or Meta", m)
		}
		c.mods = append(c.mods, m)
		c.modBits |= mod.bit
	}
	name := parts[len(parts)-1]
	if k, ok := named[name]; ok {
		c.key = k
	} else if len(name) >= 2 && name[0] == 'F' && isNumber(name[1:]) {
		n := 0
		fmt.Sscanf(name[1:], "%d", &n)
		if n < 1 || n > 12 {
			return chord{}, fmt.Errorf("function keys are F1-F12")
		}
		c.key = key{name, name, 111 + n, ""}
	} else if utf8.RuneCountInString(name) == 1 {
		r, _ := utf8.DecodeRuneInString(name)
		c.key = charKey(r, c.modBits&8 != 0)
	} else {
		return chord{}, fmt.Errorf("unknown key %q", name)
	}
	// A chord with Ctrl, Alt or Meta types nothing: it is a shortcut.
	if c.modBits&(1|2|4) != 0 {
		c.key.Text = ""
	}
	return c, nil
}

func isNumber(s string) bool {
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return s != ""
}

func charKey(r rune, shift bool) key {
	text := string(r)
	if shift {
		text = strings.ToUpper(text)
	}
	switch {
	case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z':
		u := unicode.ToUpper(r)
		return key{text, "Key" + string(u), int(u), text}
	case r >= '0' && r <= '9':
		return key{text, "Digit" + string(r), int(r), text}
	}
	return key{text, "", 0, text}
}

// printable reports whether a chord types a character into a field.
func (c chord) printable() bool { return c.key.Text != "" && c.key.Text != "\r" }
